### PR TITLE
fix(libstore): harden Retry-After handling for defensive robustness

### PR DESCRIPTION
**Pull request by randomizedcoder**
## Summary

Suggested improvements for #15449 — these edge cases probably won't matter in practice, but we've seen enough production incidents from unlikely scenarios that it felt worth hardening the retry path, just in case.

- **Honor Retry-After above maxMs** (per RFC 7231 §7.1.3): the previous code capped server-provided `Retry-After` at `maxMs`, which only makes sense for exponential backoff. If a server says "wait 120s", we should respect that, not retry after 60s.
- **Configurable Retry-After bounds** (`http-retry-after-min`, `http-retry-after-max`): clamp server values to a sane range to defend against misbehaving servers — near-zero values could drive aggressive retry loops against an already-struggling server, and enormous values could stall Nix indefinitely.
- **Jitter on Retry-After**: use the server's value as a floor and spread retries above it, so concurrent clients hitting the same 429 don't all retry at the exact same instant.
- **Parse HTTP-date in Retry-After**: use `curl_getdate()` (already linked) to handle the date form in addition to delay-seconds.
- **Extract helper functions**: `exponentialBackoffMs`, `clampRetryAfter`, `saturatingAdd`, `jitteredDelayMs` — makes the logic self-documenting and guards against unsigned overflow.
- **Expand tests**: 10 → 30 cases covering boundary conditions, degenerate configs, overflow, and jitter behavior.

## Test plan

- [x] All 30 `computeRetryDelayMs` unit tests pass
- [x] Full `nix-store-tests` suite passes
- [ ] Reviewer: verify the new settings (`http-retry-after-min`, `http-retry-after-max`) make sense as defaults (1s floor, 10min cap)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/libstore/filetransfer.cc [hunk 1]
```diff
@@ -25,35 +25,69 @@
...
-    unsigned int capped = (unsigned int) std::min(exponential, (unsigned long) maxMs);
-
-    // Retry-After is a minimum — take max(retryAfter, capped), then re-cap for sanity
-    if (retryAfterMs)
-        capped = std::min(std::max(*retryAfterMs, capped), maxMs);
+    unsigned int backoff = exponentialBackoffMs(attempt, baseMs, maxMs);
+
+    if (retryAfterMs) {
+        unsigned int serverDelay = clampRetryAfter(*retryAfterMs, retryAfterMinMs, retryAfterMaxMs);
+        if (serverDelay > backoff) {
```
Comment (Mic92) [inline src/libstore/filetransfer.cc:77]: this also solves the bug in the original PR.

File: src/libstore/filetransfer.cc [hunk 2]
```diff
@@ -25,35 +25,69 @@
...
 
 namespace nix {
 
+namespace {
+
+unsigned int exponentialBackoffMs(unsigned int attempt, unsigned int baseMs, unsigned int maxMs)
+{
+    unsigned int shift = std::min(attempt == 0 ? 0u : attempt - 1, 30u);
+    unsigned long exponential = (unsigned long) baseMs << shift;
+    return (unsigned int) std::min(exponential, (unsigned long) maxMs);
```
Comment (Mic92) [inline src/libstore/filetransfer.cc:42]: Probably will never be a problem in practice, because you would stall forever, but feels safer to have the same bugs on all platforms (32bit/64bit):

```suggestion
    uint64_t exponential = (unsigned long) baseMs << shift;
    return (unsigned int) std::min(exponential, (uint64_t) maxMs);
```

**Comment by Mic92** [comment]
@randomizedcoder can you carry over some of the code review from the original PR https://github.com/NixOS/nix/pull/15449 into this PR?

**Comment by randomizedcoder** [comment]
@Mic92 

I tried to raise this pull request against the "http-retry" branch.  Did I stuff that up?
image

**Comment by lovesegfault** [comment]
i've picked the changes i agree with from this branch into mine :)

**Comment by Mic92** [comment]
> @Mic92 
> 
> I tried to raise this pull request against the "http-retry" branch.  Did I stuff that up?
> image
> 

No I didn't saw that branch. I thought it targets main
